crypto/evp/evp_enc.c

```diff
@@ -336,6 +336,7 @@ static int evp_cipher_init_internal(EVP_CIPHER_CTX *ctx,
 
         case EVP_CIPH_CFB_MODE:
         case EVP_CIPH_OFB_MODE:
+        case EVP_CIPH_XTS_MODE:
```

Comment:
This would imply that up till this point XTS has not been used in an engine?
Why does it need to set num=0?
Are you using EVP_CIPHER_CTX_get_num() and EVP_CIPHER_CTX_set_num()

Reply:
In the ltq-deu driver I have worked on, from the kernels /crypto/testmgr perspective, calling XTS is the same thing as calling CFB or OFB, you have to provide an iv, a key (but double size). XTS has the difference that it needs a minimum of 16 bytes but allows any input length (if its not block size, cipher text stealing is used). But since in the kernel crypto mailing list there is no clear view of what the blocksize parameter for an algorithm actually means - there is no way to tell the kernel crypto API an algorithm has a minimum input length, so they used blocksize to ensure that, although like OFB or CFB it would be more logical to set blocksize to 1 and have a minimum input size.
How does an OFB or CFB caller use the get_num() and set_num() methods and in what context?
However, I have to admit, that I just assumed, that XTS has the same requirements as OFB and CFB and have placed it there. And I have tested the change with openssl speed only, which is what I have used and which I would like to have support for.

Reply:
Looking at /crypto/evp/e_aes.c get_num() and set_num() are not used by XTS. What would be the correct place then to squeeze in EVP_CIPH_XTS_MODE?

Thanks.

Reply:
The internal e_aes.c XTS  table sets the 'EVP_CIPH_CUSTOM_IV' flag - which means this whole 'if' block is skipped normally. So setting that flag may be what you want to do.. (i.e. the code block does nothing for XTS).

Reply:
This seems to be [already set](https://github.com/openssl/openssl/blob/69807ab8ed5d5b3b0bb59f418cb3a7658de005a3/crypto/evp/e_aes.c#L3261), any ideas why it is not honored and I get EVP_CipherInit_ex failure without adding case EVP_CIPH_XTS_MODE?

Reply:
Or would have the caller (openssl speed component) have to set the 'EVP_CIPH_CUSTOM_IV' flag?

Reply:
Not sure.. Are you sure it uses e_aes.c and not something else?

Reply:
@slontis The calling chain looks like openssl speed -evp calls the EVP_ functions which when any engine is involved branch to e_devcrypto, but the [flags in engines/e_devcrypto.c](https://github.com/openssl/openssl/blob/8302949262fc8f6f2639df6f0dc16b477d0fb9b0/engines/e_devcrypto.c#L122) are not set for any of the algorithms and therefor the 'legacy:' path is taken.
Question seems to be, add this single line for XTS too or add the 'EVP_CIPH_CUSTOM_IV' flag to/change XTS or all other algorithms for consistency in e_devcrypto.c?

Reply:
@slontis I have reverted the change, added your suggested changes (add the 'EVP_CIPH_CUSTOM_IV' flag) and tested it successfully.